### PR TITLE
Finished pipeline for Hopper-v2 using keras

### PR DESCRIPTION
**Pull request by mtrazzi**

### CODE REVIEW

File: quantilizer.py [hunk 1]
```diff
@@ -8,122 +8,225 @@
...
+from keras.layers import Dense, Dropout, Activation
+from keras.optimizers import SGD
+from keras import callbacks
+from datetime import datetime
+import time
+from wrappers import RobustRewardEnv
+import os
 
 def traj_segment_generator(pi, env, horizon, play=False):
+
```
Comment (RyanCarey) [inline quantilizer.py]: I was using four spaces for a tab in all of my code, so ideally we would keep that consistent.

File: quantilizer.py [hunk 2]
```diff
@@ -8,122 +8,225 @@
...
-
-    # plot
-	xticks = ["imitation"] + [str(i) for i in qs[1:]] + ["Deep Q"]
-	import ipdb; ipdb.set_trace()
-	graph_one(true_rewards, proxy_rewards, qs)
-	#graph_two(np.linspace(0,1,len(true_rewards)), true_rewards, proxy_rewards, xticks)
+	return trained_models
+
+def load_models(weights_files_list, env_name):
+	obs_dim, acs_dim = (2, 1) if env_name == 'MountainCar-v0' else (11, 27)
```
Comment (RyanCarey) [inline quantilizer.py]: you can get `obs_dim` and `acs_dim` from the env object. It's like env.actions.action_shape or something.

File: quantilizer.py [hunk 3]
```diff
@@ -8,122 +8,225 @@
...
+		models_list.append(model)
+	return models_list
+
+def extract_softmax(action):
+	"""
+	selects action with maximum probability and 
+	transforms it to correct format (inverse process as in process_labels)
+	"""
+
+	encoding = np.argmax(action) # encoding of action in 0..26
```
Comment (RyanCarey) [inline quantilizer.py]: It's worth considering whether to softmax instead of argmaxing the distribution, because maybe the most probably action is very boring -- standing still. Also, an action may be taken with ~40% probability for ten steps in a row, and be an important correction, but it will not be selected by an argmax agent.

File: quantilizer.py [hunk 4]
```diff
@@ -8,122 +8,225 @@
...
 	parser = argparse.ArgumentParser()
 	parser.add_argument("--reg", action="store", default=1e-3, type=float)
 	parser.add_argument("--hidden_size", action="store", default=20, type=int)
-	parser.add_argument("--data_set_path", action="store", default='log/MountainCar-v0/ryan.npz', type=str)	
+	parser.add_argument("--dataset_name", action="store", default='ryan', type=str)
+	parser.add_argument("--env_name", action="store", default="Hopper-v2", type=str)
+	parser.add_argument("--mode", action="store", default="train", type=str)
 	args = parser.parse_args()
-	main(reg=args.reg, hidden_size=args.hidden_size, human_dataset=args.data_set_path)
+	if (args.mode == "train"):
```
Comment (RyanCarey) [inline quantilizer.py]: makes sense

File: scripts/train_test_plot.sh [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Comment (RyanCarey) [inline scripts/train_test_plot.sh]: I'm not sure this particular file increases usability much. My personal preference is to keep everything in pure python as far as we can, and just to provide a command line argument in the README. But I'm happy with keeping this here.

File: wrappers.py [hunk 6]
```diff
@@ -46,7 +47,7 @@ def proxy_reward(self, reward, obs, done):
             return obs[0]
         elif self.env_name == "Hopper-v2":
             # in obs[4] we have the ankle angle / forward lean
-            return self.running_mean(obs[4]).mean if done else 0
+            return self.running_mean(obs[4]).mean() if done else 0
```
Comment (RyanCarey) [inline wrappers.py]: no-wonder the proxy reward was zero.

File: quantilizer.py [hunk 7]
```diff
@@ -8,122 +8,225 @@
...
+	return ob_list, ac_list, new_list, proxy_rew_list, true_rew_list
+
+def mlp_classification(input_dim, output_size, hidden_size=20):
+	model = Sequential()
+	model.add(Dense(hidden_size, activation='relu', input_dim=input_dim))
+	model.add(Dropout(0.5))
+	model.add(Dense(hidden_size, activation='relu'))
+	model.add(Dropout(0.5))
+	model.add(Dense(output_size, activation='softmax'))
+	sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
```
[the remark this answers is not in the document]
Comment (mtrazzi) [reply quantilizer.py]: Makes a lot of sense.

File: quantilizer.py (hunk 1 above)
Comment (mtrazzi) [reply quantilizer.py]: My preferences in VSCode and vim are "insert 4 spaces when tab". Don't know why but now there seems to be tabs everywhere.

File: quantilizer.py (hunk 2 above)
Comment (mtrazzi) [reply quantilizer.py]: Oh, right. I kind of "hardcoded" the `acs_dim` to be `27` for my one_hot of size 27 though. Makes sense to make it more modular.

File: quantilizer.py (hunk 3 above)
Comment (mtrazzi) [reply quantilizer.py]: Yep, will try sampling uniformly from the softmax.

File: scripts/train_test_plot.sh (hunk 5 above)
Comment (mtrazzi) [reply scripts/train_test_plot.sh]: The use case was when we need to retrain large datasets, re-doing training + testing + plotting (it's very easy to just do `sh train_test_plot.sh` "michael ryan jdonati0904 ryan0904" testplot`.

I agree that this could just be a section from the README. This file is essentially me running things on my terminal and thinking "Ryan might be interested in knowing those commands if he needs to use my code, even if it's just to know how to use the python files."

File: wrappers.py (hunk 6 above)
Comment (mtrazzi) [reply wrappers.py]: the proxy_reward was (before this modification) a numpy_object_something, and is now a float. It didn't actually run before this modification, so this was not causing anything to be zero.
